lib/semantics/mod-file.cc

```diff
@@ -590,53 +608,65 @@ std::ostream &PutLower(std::ostream &os, const std::string &str) {
   return os;
 }
 
-// Write the module file at path, prepending header. Return false on error.
-static bool WriteFile(const std::string &path, std::string &&contents) {
-  std::fstream stream;
-  auto header{GetHeader(contents)};
-  auto size{GetFileSize(path)};
-  if (size == header.size() + 1 + contents.size()) {
-    // file exists and has the right size, check the contents
-    stream.open(path, std::ios::in | std::ios::out);
-    if (FileContentsMatch(stream, header, contents)) {
-      return true;
+// Write the module file at path, prepending header. If an error occurs,
+// return an error message, otherwise nullptr.
+static const char *WriteFile(
+    const std::string &path, const std::string &contents) {
+  std::FILE *fp{std::fopen(path.c_str(), "a+")};
+  if (!fp) {
+    return std::strerror(errno);
+  }
+  if (flock(fileno(fp), LOCK_EX | LOCK_NB) != 0) {
+    fclose(fp);
+    return "Another compilation is currently using it";
+  }
+  std::size_t size = std::ftell(fp);
+  auto header{std::string{modHeader.bom} + modHeader.magic +
+      CheckSum(contents) + modHeader.terminator};
+  if (size == header.size() + contents.size()) {
+    if (FileContentsMatch(fp, header, contents)) {
+      fclose(fp);
+      return nullptr;
     }
-    stream.seekp(0);
+  }
+  std::rewind(fp);
+  ftruncate(fileno(fp), 0);
```

Comment:
I was playing around with these changes in an unrelated place, and I compiled mod-file.cc with g++ 8.3.0, and I got this message:
```
../lib/semantics/mod-file.cc:633:12: error: ignoring return value of â€˜int ftruncate(int, __off_t)â€™, declared with attribute warn_unused_result [-Werror=unused-result]
```
For some reason, this warning doesn't appear in your version of the file.  If I put in a check that the return value of ```ftruncate()==0```, the warning goes away.

Reply:
Checking the return value seems like a good idea, so I've made that change.